blockchain/chain.go

```diff
@@ -1949,6 +1949,88 @@ func (b *BlockChain) InvalidateBlock(hash *chainhash.Hash) error {
 	return err
 }
 
+// ReconsiderBlock reconsiders the validity of the block with the given hash.
+//
+// This function is safe for concurrent access.
+func (b *BlockChain) ReconsiderBlock(hash *chainhash.Hash) error {
+	b.chainLock.Lock()
+	defer b.chainLock.Unlock()
+
+	node := b.index.LookupNode(hash)
+	if node == nil {
+		// Return an error if the block doesn't exist.
+		return fmt.Errorf("Requested block hash of %s is not found "+
+			"and thus cannot be reconsidered.", hash)
+	}
+
+	// Nothing to do if the given block is already valid.
+	if node.status.KnownValid() {
+		return nil
+	}
+
+	// Clear the status of the block being reconsidered.
+	b.index.UnsetStatusFlags(node, statusInvalidAncestor)
+	b.index.UnsetStatusFlags(node, statusValidateFailed)
+
+	// Grab all the tips.
+	tips := b.index.InactiveTips(b.bestChain)
+	tips = append(tips, b.bestChain.Tip())
+
+	// Go through all the tips and unset the status for all the descendents of the
+	// block being reconsidered.
+	var reconsiderTip *blockNode
+	for _, tip := range tips {
+		// Continue if the given inactive tip is not a descendant of the block
+		// being invalidated.
+		if !tip.IsAncestor(node) {
+			// Set as the reconsider tip if the block node being reconsidered
+			// is a tip.
+			if tip == node {
+				reconsiderTip = node
+			}
+			continue
+		}
+
+		// Mark the current tip as the tip being reconsidered.
+		reconsiderTip = tip
+
+		// Unset the status of all the parents up until it reaches the block
+		// being reconsidered.
+		for n := tip; n != nil && n != node; n = n.parent {
+			b.index.UnsetStatusFlags(n, statusInvalidAncestor)
+		}
```

Comment:
Can't we break after this?

Also isn't it possible that the block we want to reconsider is a descendent of _several_ of the inactive chain tips? If this is the case, then we'll continually re-assign `reconsiderTip` for each of them. 

Consider a scenario where we had 3 inactive chain tips `{A, B, C}`, block `D` that we want to reconsider is a descendent of all 3 chain tips, but they later on fork further down the line produce distinct chain tips (that we thought were invalid in the past).

Reply:
Actually on second thought, block hash chains are unique, so if `D` is a descendent of all 3 chain tips, then the chain tips should themselves be identical.